File: code/java/188.java
class Solution {
    public int maxProfitForkDeals(int k,int[] prices){
        //需要多一次来表示初始状态
        int[] dp_i_0 = new int[k+1];
        int[] dp_i_1 = new int[k+1];
        for(int i=0;i<=k;i++) dp_i_1[i] = -Integer.MAX_VALUE;
        for(int price:prices){
            for(int i=1;i<=k;i++){
                dp_i_0[i] = Math.max(dp_i_0[i],dp_i_1[i]+price);
                dp_i_1[i] = Math.max(dp_i_1[i],dp_i_0[i-1]-price);
            }
        }
        return dp_i_0[k];
    }

    public int maxProfitForInfinityDeals(int[] prices){
        int dp_i_0 = 0;
        int dp_i_1 = -Integer.MAX_VALUE;
        int temp = -1;
        for(int price:prices){
            temp = dp_i_0;
            dp_i_0 = Math.max(dp_i_0,dp_i_1+price);
            dp_i_1 = Math.max(dp_i_1,temp-price);
        }
        return dp_i_0;
    }

    public int maxProfit(int k, int[] prices) {
        int length = prices.length;
        if(k>length/2) return maxProfitForInfinityDeals(prices);
        else return maxProfitForkDeals(k,prices);
    }
}

File: code/python/188.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        def maxProfitForK(prices,k):
            dp_i_0 = [0]*(k+1)
            dp_i_1 = [float('-inf')]*(k+1)
            for price in prices:
                for i in range(1,k+1):
                    dp_i_0[i] = max(dp_i_0[i],dp_i_1[i]+price)
                    dp_i_1[i] = max(dp_i_1[i],dp_i_0[i-1]-price)
            return dp_i_0[-1]
        def maxProfitForUnlimit(prices):
            dp_i_0 = 0
            dp_i_1 = float('-inf')
            for price in prices:
                temp = dp_i_0
                dp_i_0 = max(dp_i_0,dp_i_1+price)
                dp_i_1 = max(dp_i_1,temp-price)
            return dp_i_0
        if k>=len(prices)//2:
            return maxProfitForUnlimit(prices)
        else:
            return maxProfitForK(prices,k)